cmd/ttn-lw-cli/commands/simulate.go

```diff
@@ -0,0 +1,586 @@
+	FRMPayload  []byte          `protobuf:"bytes,12,opt,name=frm_payload,json=frmPayload,proto3" json:"frm_payload,omitempty"`
+	ConfFCnt    uint32          `protobuf:"varint,13,opt,name=conf_f_cnt,json=confFCnt,proto3" json:"conf_f_cnt,omitempty"`
+	TxDRIdx     uint32          `protobuf:"varint,14,opt,name=tx_dr_idx,json=txDRIdx,proto3" json:"tx_dr_idx,omitempty"`
+	TxChIdx     uint32          `protobuf:"varint,15,opt,name=tx_ch_idx,json=txChIdx,proto3" json:"tx_ch_idx,omitempty"`
+}
+
+var (
+	simulateUplinkFlags      = util.FieldFlags(&simulateUplinkParams{})
+	simulateJoinRequestFlags = util.FieldFlags(&simulateJoinRequestParams{})
+	simulateDataUplinkFlags  = util.FieldFlags(&simulateDataUplinkParams{})
+)
+
+func simulateDownlinkFlags() *pflag.FlagSet {
+	flagSet := &pflag.FlagSet{}
+	flagSet.Duration("timeout", 20*time.Second, "how long to wait for downlinks")
+	flagSet.Int("downlinks", 1, "how many downlinks to expect")
+	return flagSet
+}
+
+func simulate(cmd *cobra.Command, forUp func(*ttnpb.UplinkMessage) error, forDown func(*ttnpb.DownlinkMessage) error) error {
+	gtwID, err := getGatewayID(cmd.Flags(), nil, true)
+	if err != nil {
+		return err
+	}
+
+	var uplinkParams simulateUplinkParams
+	if err := util.SetFields(&uplinkParams, simulateUplinkFlags); err != nil {
+		return err
+	}
+
+	uplinkParams.setDefaults()
+
+	upMsg := &ttnpb.UplinkMessage{
+		Settings: ttnpb.TxSettings{
+			DataRate: ttnpb.DataRate{
+				Modulation: &ttnpb.DataRate_LoRa{
+					LoRa: &ttnpb.LoRaDataRate{
+						Bandwidth:       uplinkParams.Bandwidth,
+						SpreadingFactor: uplinkParams.SpreadingFactor,
+					},
+				},
+			},
+			CodingRate: "4/5",
+			Frequency:  uplinkParams.Frequency,
+			Timestamp:  uplinkParams.Timestamp,
+			Time:       uplinkParams.Time,
+		},
+		RxMetadata: []*ttnpb.RxMetadata{
+			{
+				GatewayIdentifiers: *gtwID,
+				Time:               uplinkParams.Time,
+				Timestamp:          uplinkParams.Timestamp,
+				RSSI:               uplinkParams.RSSI,
+				SNR:                uplinkParams.SNR,
+			},
+		},
+	}
+
+	if err = forUp(upMsg); err != nil {
+		return err
+	}
+
+	gs, err := api.Dial(ctx, config.GatewayServerGRPCAddress)
+	if err != nil {
+		return err
+	}
+	timeout, _ := cmd.Flags().GetDuration("timeout")
+	linkCtx, cancel := context.WithTimeout(ctx, timeout)
+	defer cancel()
+	linkCtx = rpcmetadata.MD{ID: gtwID.GatewayID}.ToOutgoingContext(linkCtx)
+	link, err := ttnpb.NewGtwGsClient(gs).LinkGateway(linkCtx)
+	if err != nil {
+		return err
+	}
+
+	// Send dummy up to start stream:
+	if err = link.Send(&ttnpb.GatewayUp{}); err != nil {
+		return err
+	}
+
+	sendTime := time.Now()
+	if err = link.Send(&ttnpb.GatewayUp{UplinkMessages: []*ttnpb.UplinkMessage{upMsg}}); err != nil {
+		return err
+	}
+
+	logger.Info("Sent uplink")
+
+	expect, _ := cmd.Flags().GetInt("downlinks")
+	for i := 0; i < expect; i++ {
+		down, err := link.Recv()
```

Comment:
I guess concurrency is for the future or is it not needed at all? The caller can optionally set the Uplink timestamp so Downlinks can arrive out of sequence right?

Reply:
With the current simulation we don't expect any downlinks before we send the uplink. Also, we don't (yet) support sending multiple uplinks, that's out of scope for these simulation commands.